ambry-commons/src/main/java/com/github/ambry/compression/BaseCompression.java

```diff
@@ -432,28 +245,35 @@ public int decompress(ByteBuffer compressedBuffer, ByteBuffer decompressedBuffer
     Utils.checkNotNullOrEmpty(compressedBuffer, "compressedBuffer cannot be null or empty.");
     Utils.checkNotNullOrEmpty(decompressedBuffer, "decompressedBuffer cannot be null or empty.");
 
-    // Save the position.  Use position instead of mark just in case decompressNative() uses mark().
+    // Check compressed buffer size - must be bigger than header size without algorithm name.
+    if (compressedBuffer.remaining() < VERSION_AND_ORIGINAL_SIZE_SIZE + ALGORITHM_NAME_LENGTH_SIZE) {
+      throw new IllegalArgumentException("compressedBuffer size of " + compressedBuffer.remaining() + " bytes is too small.");
+    }
+
+    // Check compressed buffer size - must be bigger than header size with algorithm name.
     int compressedBufferPosition = compressedBuffer.position();
-    int decompressedBufferPosition = decompressedBuffer.position();
-    try {
-      // Get the algorithm name and advance compressedBuffer position to the source data size.
-      compressedBuffer.get();  // skip the version.
-      int algorithmNameLength = compressedBuffer.get();  // Get the data source size.
-      compressedBuffer.position(compressedBuffer.position() + algorithmNameLength);
+    int algorithmNameLength = compressedBuffer.get(compressedBufferPosition + VERSION_SIZE);
+    int headerSize = VERSION_AND_ORIGINAL_SIZE_SIZE + ALGORITHM_NAME_LENGTH_SIZE + algorithmNameLength;
```

Comment:
Too many line just to index a known offset. Why can't we just directly index the buffer at the known offset ?
```
    if (compressedBuffer.remaining() < MIN_COMPRESSED_BUFFER_SIZE + compressedBuffer.get(ALGO_NAME_LEN_FIELD_INDEX)) {
      throw new IllegalArgumentException("compressedBuffer is too small and does not contain original data size.");
    }
```

Reply:
Is the `position` not always 0 ?